server/src/instant/lib/ring/websocket.clj

```diff
@@ -187,32 +191,34 @@
 
 (defn send-json!
   "Serializes `obj` to json, and sends over a websocket."
-  [app-id obj {:keys [websocket-stub undertow-websocket send-lock]}]
+  [app-id obj {:keys [websocket-stub undertow-websocket ^ReentrantLock send-lock]}]
   ;; Websockets/sendText _should_ be thread-safe
   ;; But, t becomes thread-unsafe when we use per-message-deflate
   ;; Using a `send-lock` to make `send-json!` thread-safe
   (if websocket-stub
     (websocket-stub obj)
     (let [obj-json (->json obj)
-          p (promise)
-          _ (try
-              (tracer/with-span! {:name "ws/send-json!"
-                                  :attributes {:size (count obj-json)
-                                               :app-id app-id
-                                               :send-lock.queue-length (.getQueueLength send-lock)
-                                               :send-lock.is-locked (.isLocked send-lock)
-                                               :send-lock.held-by-current-thread (.isHeldByCurrentThread send-lock)}}
-                (.lock send-lock)
-                (WebSockets/sendText
-                 ^String obj-json
-                 ^WebSocketChannel undertow-websocket
-                 (proxy [WebSocketCallback] []
-                   (complete [ws-conn context]
-                     (deliver p nil))
-                   (onError [ws-conn context throwable]
-                     (deliver p throwable)))))
-              (finally
-                (.unlock send-lock)))
-          ret @p]
-      (when (instance? Throwable ret)
-        (throw ret)))))
+          p (promise)]
+      (tracer/with-span! {:name "ws/send-json!"
```

Comment:
Tracing the full time it takes to send the json to the client. Looking at honeycomb, the send lock is almost never locked when we get here.

Reply:
Ah, perhaps it's because with grouped-queue, many of our messages are serialized per session. I wonder if we get something like set-presence _and_ transact at the same time, if the lock will start to come in effect

Reply:
I think the reason we rarely wait on the lock is that calling `Websockets/sendText` is very quick--undertow is probably just putting a message onto a queue. All of the waiting happens between when we call `sendText` and when the `complete` callback is called.